noobaa/noobaa-core
# S3 Versioning

![Tests](https://gdurl.com/jz2I)

### Explain the changes
#### 1. New Features / Capabilities (Sync with Liran):
- S3 versioning by @jeniawhite 

#### 2. Existing Behavior Changes (Sync with Liran):
- Many of the existing code paths of get/put/delete object were updated to support versioning, even without versioning enabled.
- Performance of adding objects to the system might change due to added indexes.

#### 3. Other Changes:
- Removed Cloud-sync code.

### Issues: Fixed #xxx / Gap #xxx
- None?

### Testing Instructions:
- Enable bucket versioning (no frontend yet)
- Suspend bucket versioning (no frontend yet)
- List-Object-Versions (with awscli/cyberduck)
- Get/Put/Delete objects with versioning Disabled, Enabled, Suspended.
- There are multiple corner cases of the S3 api behavior.
- Check that blob is not affected.

### Reminders:
- Create a new test - the first is the hardest
- User Experience is top priority
- Design for failure
- Keep it simple
- `npm test`
- Imagine a support call - Add diagnostics info, phonehome info, activity log events, external syslog
- Upgrade - DB schema, server platform, agents install, npm packages
